### ISSUE TITLE
Does not meet the witx spec while executing the given wasm file

### ISSUE THREAD

**Issue by luxinyi0105** [WasmEdge/WasmEdge#2694]
## Description

WasmEdge may not be compliant with the witx spec while executing the given test case.

Actually, the given test case was simply mutated a wasm file, which was obtained by compiling the C program generated with Csmith using the Emscripten compiler (Emcc). You can convert wasm file to wat format with wabt and find the mutation operation in line 9974(replace `i32.sub` with `i32.div_s`). Releated files is [here](https://github.com/luxinyi0105/wasm_runtime_test/raw/main/07-26.zip).

### Current State

```sh
$ wasmedge mutated.wasm 
checksum = DAFC642C
```

### Expected

As mentioned above, the mutation there is changing a subtraction of the stack pointer into a division of the stack pointer which means that the wasm module's in-memory shadow-stack pointer is getting corrupted. Notably what this causes is the stack to get misaligned.

[witx spec](https://github.com/WebAssembly/wasi/blob/main/legacy/tools/witx-docs.md#pointers) says `misaligned pointers ... shall trap.` It seems that WasmEdge needs to return a trap while meeting this situatuion, but actually it correctly output the checksum result and did not trigger any abnormal behavior.

## Environment

- Hardware Architecture: x86_64
- Operating system: Ubuntu 22.04
- Tools Version: wasmedge 0.13.2

**Comment by bmorphism** [comment]
alignment is hard, but eventually all shoggoths shall be trapped! this looks fun and I may attempt this if not closed by October

**Comment by q82419** [comment]
The [witx-spec](https://github.com/WebAssembly/wasi/blob/main/legacy/tools/witx-docs.md#pointers) mentioned is the pointers passed into the WASI functions.
For the WASM internal, refer to the note in [WASM spec](https://webassembly.github.io/spec/core/exec/instructions.html#memory-instructions):

```
Unaligned access violating that property is still allowed and must succeed regardless of the annotation. However, it may be substantially slower on some hardware.
```

Therefore I think the implementation of detecting the misaligned pointers should be in the WASI functions, not the memory instance.

**Comment by DhruvSinghiitmandi** [comment]
Hi @q82419  ,
I have been working on this issue for the last couple of days.
On using gdb to see the call stack . Three host functions from lib/host/wasi/wasifunc.cpp were being called.
I have added alignment checks to these functions as per the [WASI spec](https://github.com/WebAssembly/WASI/blob/main/legacy/preview1/docs.md#modules). A TRAP will now be thrown if the pointer is misaligned.

Current behaviour

<img width="752" height="81" alt="Image" src="https://github.com/user-attachments/assets/b760863e-b889-4873-9a55-c8215a9b8a4e" />

Changed Behaviour(Verified through gdb that there is a error thrown at after the breakpoint of WasiArgsGet

<img width="926" height="46" alt="Image" src="https://github.com/user-attachments/assets/2e5863db-d285-4d98-a498-9a1dc1196866" />

But currently I am not able to see any error message for the TRAP, also the exit code after echo $? is 0 (Compared with an example that raises a TRAP)

<img width="737" height="79" alt="Image" src="https://github.com/user-attachments/assets/2b2114c6-a257-4871-b024-a9e8c1417bff" />

<img width="753" height="63" alt="Image" src="https://github.com/user-attachments/assets/703e4334-0a60-421d-98f8-0d8aa67a4311" />



How should I proceed ? should I raise A PR?

**Comment by AbhinavJha1023** [comment]
Hi @q82419,
I've been following this issue and the recent discussions with a lot of interest. The behavior around misaligned pointers not triggering traps in WASI host functions caught my attention, especially considering what the spec expects.

The mutated WASM file corrupts the shadow stack pointer using a division instead of subtraction, which results in a misaligned pointer. While this is technically allowed by the core WebAssembly spec, I understand that **WASI host functions** need to enforce stricter alignment, and misaligned pointers passed to them should **trap**, according to the WITX spec.

Right now, though, WasmEdge executes the file and returns a checksum, so the trap never happens — which seems like something worth fixing.

I've started looking into the relevant host functions inside `wasifunc.cpp`, especially things like `args_get` and `environ_get`, and I'm working on adding alignment checks like this:

if ((reinterpret_cast<uintptr_t>(pointer) % alignof(T)) != 0) {
    return __WASI_ERRNO_INVAL;
}

This would help catch unaligned pointers early and return a proper error.

1. **Should this return an error or throw a trap?**
   Is returning something like `__WASI_ERRNO_INVAL` enough? Or should I raise a trap using something like `throw WasmEdge::ErrCode::Unreachable`?

2. **About the exit code:**
   I noticed that even with the misalignment check, the program still exits with code `0`. Is there a recommended way to make sure a trap or error actually leads to a non-zero exit code? That would really help with testing.

3. **Would you prefer a small PR first?**
   I can make a focused PR for just one function (like `args_get`) to start with, so it’s easy to review and discuss. Happy to expand from there.

Thanks a lot for putting together this project — it’s a really interesting area and I’d love to contribute in any way I can.

**Comment by q82419** [comment]
Hi @DhruvSinghiitmandi  @AbhinavJha1023 

This issue has been packed into the LFX mentorship entry. Please check the issue #4280 .
If you interest in this, please feel free to apply the program. Thanks!

**Comment by leetcodez** [comment]
Hey @q82419, I've been digging into this one and tracing the execution path locally. It looks like WasmEdge correctly outputting the checksum without trapping in the core execution is actually perfectly compliant with the W3C Core Spec(unaligned memory access are required to succeed there). The real gap seems to be at the WASI boundary.

The mutated pointer from the i32.div_s shadow stack corruption is slipping through host functions like args_get or environ_get inside wasifunc.cpp. If we just return` __WASI_ERRNO_INVAL` there, the WASM guest can just ignore the errno and exit gracefully with code 0. To enforce the trap properly and get a non-zero exit code like the WITX spec demands, we probably need to return a fatal expected error directly from the host function; something like `spdlog::error`(`ErrCode::Value::TrapUnalignedAccess`) to violently halt the VM.

**Comment by parthdagia05** [comment]
@hydai is this issue open for contirbutors?

**Comment by hydai** [comment]
Nope, this is included in #4820, which belongs to the mentee for LFX 2026 term 2.
